Sources/SwiftASN1/Basic ASN1 Types/PEMDocument.swift

```diff
@@ -11,16 +11,149 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 //===----------------------------------------------------------------------===//
+
 #if canImport(Foundation)
 import Foundation
 
+/// Defines a type that can be serialized in PEM-encoded form.
+///
+/// Users implementing this type are expected to just provide the ``defaultPEMDiscriminator``
+///
+/// A PEM `String` can be serialized by constructing a ``PEMDocument`` by calling ``PEMSerializable/serializeAsPEM()`` and then accessing the ``PEMDocument/pemString`` preropty.
+public protocol PEMSerializable: DERSerializable {
+    /// The PEM discriminator identifying this object type.
+    ///
+    /// The PEM discriminator is in the first line of a PEM string after `BEGIN` and at the end of the string after `END` e.g.
+    /// ```
+    /// -----BEGIN defaultPEMDiscriminator-----
+    /// <base 64 DER representation of this object>
+    /// -----END defaultPEMDiscriminator-----
+    /// ```
+    static var defaultPEMDiscriminator: String { get }
+    
+    func serializeAsPEM(discriminator: String) throws -> PEMDocument
+}
+
+/// Defines a type that can be parsed from a PEM-encoded form.
+///
+/// Users implementing this type are expected to just provide the ``defaultPEMDiscriminator``.
+///
+/// Objects that are ``PEMParseable`` can be construct from a PEM `String` through ``PEMParseable/init(pemEncoded:)``.
+public protocol PEMParseable: DERParseable {
+    /// The PEM discriminator identifying this object type.
+    ///
+    /// The PEM discriminator is in the first line of a PEM string after `BEGIN` and at the end of the string after `END` e.g.
+    /// ```
+    /// -----BEGIN defaultPEMDiscriminator-----
+    /// <base 64 DER representation of this object>
+    /// -----END defaultPEMDiscriminator-----
+    /// ```
+    static var defaultPEMDiscriminator: String { get }
+    
+    
+    /// The PEM discriminator allowed to be present when parsing a PEM string.
+    ///
+    /// The PEM discriminator is in the first line of a PEM string after `BEGIN` and at the end of the string after `END` e.g.
+    /// ```
+    /// -----BEGIN defaultPEMDiscriminator-----
+    /// <base 64 DER representation of this object>
+    /// -----END defaultPEMDiscriminator-----
+    /// ```
+    static var allowedPEMDiscriminators: Set<String> { get }
```

Comment:
I think this might be excessive: we can probably just allow the user code to handle this check in `init(pemDocument:)`.

Reply:
removed